compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjFifos.cpp

```diff
@@ -13,6 +13,150 @@
 
 namespace mlir::iree_compiler::AMDAIE {
 
+/// Utility to compute the product of the sizes after each dimension:
+/// productSizes[idx] will contain the product of all sizes in [idx + 1,
+/// sizes.size() - 1].
+SmallVector<int64_t> getProductSizesAfterIndex(ArrayRef<int64_t> sizes) {
+  SmallVector<int64_t> productSizes;
+  productSizes.reserve(sizes.size());
+  int64_t productSize{1};
+  for (auto size : llvm::reverse(sizes)) {
+    productSize *= size;
+    productSizes.push_back(productSize);
+  }
+  std::reverse(productSizes.begin(), productSizes.end());
+  return productSizes;
+}
+
+/// Utility struct to represent DMA split information.
+struct DmaSplitInfo {
+  size_t sourceSplitDim;
+  size_t targetSplitDim;
+};
+
+using DmaObjFifoPairT =
+    std::pair<AMDAIE::DmaCpyNdOp, AMDAIE::LogicalObjectFifoFromMemrefOp>;
+
+/// Find the logical objectFifo and DMA source/target splitting dimensions for
+/// each DMA and objectFifo pair.
+///
+/// Each pair is handled in the following way:
+/// First, compute the objectFifo splitting dimension as the last non-unit shape
+/// dimension. Afterwards, depending on which logical objectFifo is being
+/// split on, find the outermost dimension in either the source or
+/// target access pattern that has:
+/// - stride == sizeAfterSplit
+/// - size != 1
+/// This is the splitting dimension to be used on the respective side of the DMA
+/// operation. Then, calculate the product size of that side of the DMA
+/// operation after the splitting dimension and use it to calculate the
+/// splitting dimension on the other side as the first dimension from the back
+/// that has product size larger than the other side's product size after
+/// splitting because that's the number of elements that should be
+/// produced/consumed on the respective sides before splitting.
+LogicalResult collectSplittingDims(
+    const SmallVector<DmaObjFifoPairT> &dmaObjFifoPairs,
+    DenseMap<AMDAIE::DmaCpyNdOp, DmaSplitInfo> &dmaSplitInfoMap,
+    DenseMap<AMDAIE::LogicalObjectFifoFromMemrefOp, size_t>
+        &objFifoSplitDimMap) {
+  for (auto [dmaOp, objFifo] : dmaObjFifoPairs) {
+    LLVM_DEBUG(llvm::dbgs() << "dmaOp: " << dmaOp << "\n");
+    LLVM_DEBUG(llvm::dbgs() << "objFifo: " << objFifo << "\n");
+    ArrayRef<int64_t> memrefShape = objFifo.getMemrefType().getShape();
+    if (llvm::any_of(memrefShape, [](int64_t size) { return size < 0; })) {
+      return objFifo.emitOpError()
+             << "can't find a valid split dimension for dynamic sizes memref";
+    }
+    size_t objFifoSplitDim{0};
```

Comment:
auto iter = std::find_if(memrefShape.begin(), memrefShape.end(),
                         [](int64_t size) { return size > 1; });
auto objFifoSplitDim = std::distance(memrefShape.begin(), iter);

I guess an edge case is when they're all 1, not sure what the correct behaviour is then

Reply:
Thanks, using this now. The edge case of all ones should not split, so added a check for this at this place.